lib/snowflex/transport/http/type.ex

```diff
@@ -0,0 +1,189 @@
+  @typedoc "Output value."
+  @type return_value ::
+          bitstring()
+          | integer()
+          | date()
+          | datetime()
+          | Decimal.t()
+
+  @typedoc "Date as `{year, month, day}`"
+  @type date :: {1..9_999, 1..12, 1..31}
+
+  @typedoc "Time as `{hour, minute, sec, usec}`"
+  @type time :: {0..24, 0..60, 0..60, 0..999_999}
+
+  @typedoc "Datetime"
+  @type datetime :: {date(), time()}
+
+  @type encoded_value :: %{type: String.t(), value: String.t() | nil}
+
+  @typedoc "Type information for encoding/decoding"
+  @type type_info :: map()
+
+  @doc """
+  Transforms input params for the REST API.
+  The REST API expects values in JSON format with type specified.
+
+  See: https://docs.snowflake.com/en/developer-guide/sql-api/submitting-requests#using-bind-variables-in-a-statement
+  """
+  @spec encode(encodeable(), term()) :: encoded_value()
+  def encode(true, _), do: %{type: "BOOLEAN", value: "true"}
+  def encode(false, _), do: %{type: "BOOLEAN", value: "false"}
+  def encode(nil, _), do: %{type: "TEXT", value: nil}
+
+  def encode(param, _) when is_integer(param),
+    do: %{type: "FIXED", value: to_string(param)}
+
+  def encode(param, _) when is_float(param),
+    do: %{type: "REAL", value: to_string(param)}
+
+  def encode(param, _) when is_atom(param) do
+    %{type: "TEXT", value: Atom.to_string(param)}
+  end
+
+  def encode(%Decimal{} = param, _) do
+    %{type: "FIXED", value: to_string(param)}
+  end
+
+  def encode(%Date{} = param, _) do
+    %{type: "TEXT", value: Date.to_iso8601(param)}
+  end
+
+  def encode(%Time{} = param, _) do
+    %{type: "TEXT", value: Time.to_iso8601(param)}
+  end
+
+  def encode(%DateTime{} = param, _) do
+    %{type: "TEXT", value: DateTime.to_iso8601(param)}
+  end
+
+  def encode(%NaiveDateTime{} = param, _) do
+    %{type: "TEXT", value: NaiveDateTime.to_iso8601(param)}
+  end
+
+  def encode(param, _) when is_binary(param) do
+    %{type: "TEXT", value: param}
+  end
+
+  def encode(value, _) do
+    raise Error.exception(
+            message: "could not parse param #{inspect(value)} of unrecognised type."
+          )
+  end
+
+  @doc """
+  Transforms REST API return values to Elixir representations.
+  """
+  @spec decode(nil, term()) :: nil
+  def decode(nil, _), do: nil
+
+  @spec decode(binary(), %{type: type_info()} | %{column: binary()}) :: return_value()
+  def decode(value, %{type: type_info}) when is_binary(value) do
+    column_type = String.downcase(Map.get(type_info, "type", ""))
+    decode_by_type(value, column_type, type_info)
+  end
+
+  # Fallback for when we don't have type information but have column name
+  def decode(value, %{column: column_name}) when is_binary(value) and is_binary(column_name) do
+    column_name = String.downcase(column_name)
+
+    if String.contains?(column_name, "id") and integer_string?(value) do
+      # If column name contains "id" and looks like an integer, convert to integer
+      String.to_integer(value)
+    else
```

Comment:
This feels unintuitive - is there a reason to include this at the adapter level? This would produce inconsistent behaviour with e.g. alphanumeric IDs, where `abc123` would stay as a string but `123456` would be converted to an integer.